packages/resolve-url-loader/docs/advanced-features.md

```diff
@@ -1 +1,219 @@
+
+* The webpack file-system is provided by the `enhanced-resolver-plugin` and does not contain `fs.existsSync`. Instead, use the `webpackExistsSync` utility function as shown.
+
+* You may set additional `options` when you configure the loader in webpack that you can then access in the generator. In this case the `attempts` could be made a configurable option.
+
+## Reference
+
+For full reference check the source code in [`lib/join-function/index.js`](../lib/join-function/index.js).
+
+The default "join" function is exported as `defaultJoin` and is equivalent to the following.
+
+```javascript
+const {
+  createJoinFunction,
+  defaultJoinGenerator,
+  defaultJoinOperation,
+  defaultJoin
+} = require('resolve-url-loader');
+
+// create a join function equivalent to "defaultJoin"
+const myJoinFn = createJoinFunction({
+  name     : 'myJoinFn',
+  scheme   : 'alstroemeria',
+  generator: defaultJoinGenerator,
+  operation: defaultJoinOperation
+});
+```
+
+The `name` is used purely for debugging purposes.
+
+The `scheme` should be a literal string of the current scheme and should match the value shown in the loader `package.json` at the time you first author your custom join function.
+
+The `generator` chooses the order of potential base paths to consider.
+
+```javascript
+generator (filename: string, uri: string, bases: {}, isAbsolute:boolean, options: {}) => Array<string> | Iterable<string> 
+```
+
+* The `filename` is the loader `resourcePath`.
+* The `uri` is the argument to the `url()` as it appears in the source file.
+* The `bases` are a hash where the keys are the sourcemap evaluation locations in the [algorithm](./how-it-works.md#algorithm) and the values are absolute paths that the sourcemap reports. These directories might not actually exist.
+* The `isAbsolute` flag indicates whether the URI is considered an absolute file or root relative path by webpack's definition. Absolute URIs are only processed if the `root` option is specified.
+* The `options` are the loader options as configured in webpack. This includes documented options as well as any you add in your configuration.
+
+The `operation` is the predicate which determines whether a base path is successful. Each may be modified or the default ones used as shown above.
+
+```javascript
+operation (filename: string, uri: string, base: string, next: function, options: {}) => string | typeof call<next>
+```
+
+* The `filename` is the loader `resourcePath`.
+* The `uri` is the argument to the `url()` as it appears in the source file.
+* The `base` is a single base path where the URI may exist. This directory might not actually exist.
+* The `next` function is only called on failure. Passing any value to `next(...)` marks that as a fallback value but only the earliest fallback is ever used. Always `return` the result of calling `next()`.
+* The `options` are the loader options as configured in webpack. This includes documented options as well as any you add in your configuration.
```

Comment:
It might just be me, but I ended up finding much more success by reading these docs in reverse. I found that seeing the api first, then reading the example, gave me a lot more context and a reference point by the time I got to the building blocks and how to section.

As a general rule I think I like when docs "zoom in" - ie start with the high level description and examples and what all the pieces look like, and then slowly dive into each individual part + give extra details + in depth descriptions

Reply:
Thats interesting. Thats how I prefer but I'd had other advice that implied I should defer the reference. 🤔

Reply:
changed the order, put an index at the top for @dcousens